Introduce make_rcp(...)

kinnewig opened the pull request:
As highlighted in Issue #16202, we want to avoid using the `operator new` directly. In Trilinos Version 14.0.0, the function `Teuchos::make_rcp()` was introduced, which does precisely that. But to ensure backward compatibility with older Trilinos Versions, I introduce `Utilities::Trilinos::make_rcp()` (as suggested in Issue #16202).
Depending on the used Trilinos version, either `Teuchos::make_rcp()` or `Utilities::Trilinos::make_rcp()` is used. In my opinion, this is the best approach to ensure compatibility with future Trilinos versions. 

An alternative to my approach would be to use `Utilities::Trilinos::make_rcp()` independent of the Trilinos version. Please let me know your thoughts!

masterleinad commented on source/base/trilinos_utilities.cc:
```diff
@@ -169,6 +169,21 @@ namespace Utilities
+#ifdef DEAL_II_TRILINOS_WITH_TPETRA
+  namespace Trilinos
+  {
+    template <class T, class... Args>
+    Teuchos::RCP<T>
+    make_rcp(Args &&...args)
+    {
+      return Teuchos::RCP<T>(new T(std::forward<Args>(args)...));
+    }
```
We could avoid a lot of code changes by always using `Utilities::Trilinos::make_rcp` and gaming it an alias for `Teuchos::rcp` in case that is available.
```suggestion
#ifdef DEAL_II_TRILINOS_VERSION_GTE(14, 0, 0)
     using Teuchos::make_rcp;
#else
    template <class T, class... Args>
    Teuchos::RCP<T>
    make_rcp(Args &&...args)
    {
      return Teuchos::RCP<T>(new T(std::forward<Args>(args)...));
    }
#endif
```

kinnewig replied:
I like that solution, it is somewhere between the both approaches I suggested.

masterleinad reviewed (commented):
I would prefer using `Utilities::Trilinos::make_rcp` in all the places unconditionally.

kinnewig commented on include/deal.II/base/trilinos_utilities.h:
```diff
@@ -180,6 +184,34 @@ namespace Utilities
+     * was introduce. If Trilinos version 14.0.0 or higher is detected,
+     * we use an alias for that function. Otherwise, we make our own version
+     * of make_rcp(). our own version of make_rcp().
+     */
```
I implemented the suggestions but am unsure how to write proper doxygen documentation in this case. This does not show up in the doxygen.

masterleinad commented on source/base/trilinos_utilities.cc:
```diff
@@ -169,6 +169,25 @@ namespace Utilities
+
+
+#ifdef DEAL_II_TRILINOS_WITH_TPETRA
+  namespace Trilinos
```
You need to add doxygen documentation for this `Trilinos` namespace to get documentation for `make_rcp`, see https://stackoverflow.com/a/35068555.

kinnewig replied:
This does not resolve that problem. 
The Doxygen settings from deal.II seem to suppress the generation of Doxygen entries for aliases.

I just pushed a fix for that problem. If that fix is okay, I will squash the commits.

masterleinad commented on doc/news/changes/minor/20231115SebastianKinnewig:
```diff
@@ -0,0 +1,7 @@
+New: Introduce a new function, Utilities::Trilinos::make_rcp(),
+to create Teuchos::RCP objects, avoiding the usage of
+'operator new'.
+Fixed: Replace 'operator new' usage with function calls of
+Utilities::Trilinos::make_rcp().
```
It wasn't really a bug, right?

kinnewig replied:
No, it was not a bug.
should I just keep that sentence in the "new" section, or should I just remove it?

masterleinad replied:
I would just remove it since the changes are not user-facing.

masterleinad commented on include/deal.II/base/trilinos_utilities.h:
```diff
@@ -180,6 +184,34 @@ namespace Utilities
+     * @note In Trilinos 14.0.0, the function
+     * <a
+     * href="https://docs.trilinos.org/dev/packages/teuchos/doc/html/namespaceTeuchos.html#a280c0ab8c9ee8d0481114d4edf5a3393">Teuchos::make_rcp()</a>
+     * was introduce. If Trilinos version 14.0.0 or higher is detected,
+     * we use an alias for that function. Otherwise, we make our own version
+     * of make_rcp().
```
```suggestion
     * was introduce.
```
I don't think we need to expose implementation details.

masterleinad commented on include/deal.II/base/trilinos_utilities.h:
```diff
@@ -180,6 +184,32 @@ namespace Utilities
+     * @note In Trilinos 14.0.0, the function
+     * <a
+     * href="https://docs.trilinos.org/dev/packages/teuchos/doc/html/namespaceTeuchos.html#a280c0ab8c9ee8d0481114d4edf5a3393">Teuchos::make_rcp()</a>
+     * was introduce.
```
```suggestion
     * was introduced.
```

masterleinad reviewed (commented):
Apart from that last typo, this looks good to me.

bangerth commented on doc/news/changes/minor/20231115SebastianKinnewig:
```diff
@@ -0,0 +1,5 @@
+to create Teuchos::RCP objects, avoiding the usage of
+'operator new'.
+<br>
+(Sebastian Kinnewig, 2023/11/15)
```
You'll need an endline at the end of this line.

kinnewig replied:
Done!

bangerth commented on include/deal.II/base/trilinos_utilities.h:
```diff
@@ -180,6 +184,32 @@ namespace Utilities
+
+
+#ifdef DEAL_II_TRILINOS_WITH_TPETRA
+  namespace Trilinos
```
We already have `namespace TrilinosWrappers`. Why not put it in there? If you did, you don't have to qualify the use of this function at all if you're inside that namespace (which should be all the places where you're using it) and in that case you can play the following game:
```
namespace TrilinosWrappers
{
#if DEAL_II_TRILINOS_VERSION_LT(too old)
   ...provide make_rcp...
#  else
    using Teuchos::make_rcp;
#endif


...some code inside namespace TrilinosWrappers:
x = make_rcp<...>(...);
```
The latter code uses `TrilinosWrappers::make_rcp()` if provided (namely, if Trilinos is too old) and falls back to `Teuchos::make_rcp` otherwise. Once we get to a point where we *require* a new enough Trilinos library, we can just remove the `#if` part.

kinnewig replied:
Sure, we can move that into the `namespace TrilinosWrappers`. But then the question arises: in which file should I place that function?

With that argument, moving that function into the `namespace TpetraWrappers` could be even more convenient since I am using that function quite heavily in that namespace.

bangerth replied:
Ah, I had not realized that `TpetraWrappers` are not in namespace `TrilinosWrappers` :-(

I guess then this is ok as is.

bangerth commented on include/deal.II/base/trilinos_utilities.h:
```diff
@@ -180,6 +184,32 @@ namespace Utilities
+     * href="https://docs.trilinos.org/dev/packages/teuchos/doc/html/namespaceTeuchos.html#a280c0ab8c9ee8d0481114d4edf5a3393">Teuchos::make_rcp()</a>
+     * was introduced.
+     */
+#  if defined DOXYGEN || !DEAL_II_TRILINOS_VERSION_GTE(14, 0, 0)
```
I'm unsure whether this is necessary, but we seem to do it as follows everywhere else:
```suggestion
#  if defined(DOXYGEN) || !DEAL_II_TRILINOS_VERSION_GTE(14, 0, 0)
```

kinnewig replied:
Both should work. But I will adapt it, so it matches the rest.

bangerth commented:
/rebuild

bangerth commented:
Perhaps one last option: I think we eventually want to get rid of this function again, once we require at least Trilinos 14. But right now it's part of the deal.II public interface, so we can't get rid of it again. We could make it non-public by putting into a namespace `Utilities::Trilinos::internal`. Thoughts?

kinnewig commented:
> Perhaps one last option: I think we eventually want to get rid of this function again, once we require at least Trilinos 14. But right now it's part of the deal.II public interface, so we can't get rid of it again. We could make it non-public by putting into a namespace `Utilities::Trilinos::internal`. Thoughts?

I guess that is a good idea, since `Teuchos::make_rcp()` should be preferred anyway.

kinnewig commented:
I just moved make_rcp into `Utilities::Trilinos::internal`

And added a note to the documentation of Utilities::Trilinos::internal::make_rcp() that Teuchos::make_rcp() should be preferred.

bangerth commented on doc/news/changes/minor/20231115SebastianKinnewig:
```diff
@@ -0,0 +1,5 @@
+New: Introduce a new function, Utilities::Trilinos::make_rcp(),
```
You'll have to change this as well.

kinnewig replied:
Thank you, I missed that one

bangerth reviewed (commented):
OK with the one change to the changelog.

bangerth commented:
Excellent, thank you!
